src/main/scala/ch/epfl/scala/bsp/Bsp.scala

```diff
@@ -0,0 +1,401 @@
+
+final case class Uri private[Uri](val s: String)
+object Uri {
+  // This is the only valid way to create a URI
+  def apply(u: URI): Uri = Uri(u.toString)
+
+  implicit val uriEncoder: RootEncoder[Uri] = new RootEncoder[Uri] {
+    override def apply(a: Uri): Json = Json.fromString(a.s)
+  }
+
+  implicit val uriDecoder: Decoder[Uri] = new Decoder[Uri] {
+    // Should we add validation here
+    override def apply(c: HCursor): Result[Uri] = {
+      c.as[String].flatMap { str =>
+        try Right(Uri(URI.create(str).toString))
+        catch {
+          case _: IllegalArgumentException | _: URISyntaxException =>
+            Left(DecodingFailure(s"String $str is not a valid URI.", c.history))
+        }
+      }
+    }
+  }
+}
+
+@JsonCodec final case class TextDocumentIdentifier(
+    uri: Uri
+)
+
+@JsonCodec final case class BuildTargetIdentifier(
+    uri: Uri
+)
+
+@JsonCodec final case class BuildTargetCapabilities(
+    canCompile: Boolean,
+    canTest: Boolean,
+    canRun: Boolean,
+)
+
+@JsonCodec final case class BuildTarget(
+    id: BuildTargetIdentifier,
+    displayName: String,
+    languageIds: List[String],
+    dependencies: BuildTargetIdentifier,
+    capabilities: BuildTargetCapabilities,
+    data: Option[Json]
+)
+
+@JsonCodec final case class BuildClientCapabilities(
+    languageIds: List[String]
+)
+
+// Notification: 'build/initialized', C -> S
+@JsonCodec final case class InitializedBuildParams()
+
+// Request: 'build/initialize', C -> S
+@JsonCodec final case class InitializeBuildParams(
+    rootUri: Uri,
+    capabilities: BuildClientCapabilities
+)
+
+@JsonCodec final case class Shutdown()
+
+@JsonCodec final case class Exit()
+
+@JsonCodec final case class CompileProvider(
+    languageIds: List[String]
+)
+
+@JsonCodec final case class TestProvider(
+    languageIds: List[String]
+)
+
+@JsonCodec final case class RunProvider(
+    languageIds: List[String]
+)
+
+@JsonCodec final case class BuildServerCapabilities(
+    compileProvider: CompileProvider,
+    testProvider: TestProvider,
+    runProvider: RunProvider,
+    providesTextDocumentBuildTargets: Boolean,
+    providesDependencySources: Boolean,
+    providesResources: Boolean,
+    providesBuildTargetChanged: Boolean
+)
+
+@JsonCodec final case class InitializeBuildResult(
+    capabilities: BuildServerCapabilities
+)
+
+sealed abstract class MessageType(val id: Int)
+object MessageType {
+  case object Error extends MessageType(1)
+  case object Warning extends MessageType(2)
+  case object Info extends MessageType(3)
+  case object Log extends MessageType(4)
+
+  implicit val messageTypeEncoder: RootEncoder[MessageType] = new RootEncoder[MessageType] {
+    override def apply(a: MessageType): Json = Json.fromInt(a.id)
+  }
+
+  implicit val messageTypeDecoder: Decoder[MessageType] = new Decoder[MessageType] {
+    override def apply(c: HCursor): Result[MessageType] = {
+      c.as[Int].flatMap {
+        case 1 => Right(Error)
```

Comment:
I think we should abstract over this boilerplate with a simple function

```scala
def enumDecoder[T](allValues: List[T], getId: T => Int, onError: Int => String): Decoder[T] = ...
def enumEncoder[T](allValues: List[T], getId: T => Int): Encoder[T] = ...

sealed trait MessageType
object MessageType {
  val all = List(Error, Warning, Info)
  implicit val decoder  = enumDecoder(all, _.id, n => s"Bad id $n")
```

That avoids the risk of doing a copy-paste mistake in the `case MAGIC_CONSTANT` pattern match.

Alternatively, we use enumeratum library in lsp4s to avoid doing this by hand

https://github.com/scalameta/lsp4s/blob/de927af3a9e9664a650b8247c9280e777388bda1/lsp4s/src/main/scala/scala/meta/lsp/Enums.scala#L5-L17

It's a relatively small dependency (48K).

Reply:
I agree this is much nicer, but let's do this after 1.0 is out. It won't change semantics and it only affects the implementation and maintenance of this tiny library.